rust/datafusion/src/optimizer/constant_folding.rs

```diff
@@ -0,0 +1,834 @@
+//! unary expression.
+
+use std::sync::Arc;
+
+use arrow::datatypes::DataType;
+
+use crate::error::Result;
+use crate::logical_plan::{DFSchemaRef, Expr, LogicalPlan, Operator};
+use crate::optimizer::optimizer::OptimizerRule;
+use crate::optimizer::utils;
+use crate::scalar::ScalarValue;
+
+/// Optimizer that simplifies comparison expressions involving boolean literals.
+///
+/// Recursively go through all expressionss and simplify the following cases:
+/// * `expr = true` and `expr != false` to `expr` when `expr` is of boolean type
+/// * `expr = false` and `expr != true` to `!expr` when `expr` is of boolean type
+/// * `true = true` and `false = false` to `true`
+/// * `false = true` and `true = false` to `false`
+/// * `!!expr` to `expr`
+/// * `expr = null` and `expr != null` to `null`
+pub struct ConstantFolding {}
+
+impl ConstantFolding {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl OptimizerRule for ConstantFolding {
+    fn optimize(&self, plan: &LogicalPlan) -> Result<LogicalPlan> {
+        match plan {
+            LogicalPlan::Filter { predicate, input } => Ok(LogicalPlan::Filter {
+                predicate: optimize_expr(predicate, &plan.all_schemas())?,
+                input: Arc::new(self.optimize(input)?),
+            }),
+            // Rest: recurse into plan, apply optimization where possible
+            LogicalPlan::Projection { .. }
+            | LogicalPlan::Aggregate { .. }
+            | LogicalPlan::Repartition { .. }
+            | LogicalPlan::CreateExternalTable { .. }
+            | LogicalPlan::Extension { .. }
+            | LogicalPlan::Sort { .. }
+            | LogicalPlan::Explain { .. }
+            | LogicalPlan::Limit { .. }
+            | LogicalPlan::Join { .. } => {
+                // apply the optimization to all inputs of the plan
+                let inputs = utils::inputs(plan);
+                let new_inputs = inputs
+                    .iter()
+                    .map(|plan| self.optimize(plan))
+                    .collect::<Result<Vec<_>>>()?;
+
+                let schemas = plan.all_schemas();
+                let expr = utils::expressions(plan)
+                    .iter()
+                    .map(|e| optimize_expr(e, &schemas))
+                    .collect::<Result<Vec<_>>>()?;
+
+                utils::from_plan(plan, &expr, &new_inputs)
+            }
+            LogicalPlan::TableScan { .. } | LogicalPlan::EmptyRelation { .. } => {
+                Ok(plan.clone())
+            }
+        }
+    }
+
+    fn name(&self) -> &str {
+        "constant_folding"
+    }
+}
+
+fn is_boolean_type(expr: &Expr, schemas: &[&DFSchemaRef]) -> bool {
+    for schema in schemas {
+        match expr.get_type(schema) {
+            Ok(dt) if dt == DataType::Boolean => {
+                return true;
+            }
+            _ => {
+                continue;
```

Comment:
You might be able to write this more concisely with something like 
```
if let Ok(DataType::Boolean) = expr.get_type(schema) {
  return true
}
```

Reply:
Good call! I have also added more comments to explain why `all_schemas` method was needed in the optimizer code.